source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,193 @@
+
+### Vdsm multipath event listener
+
+Introduce a new Vdsm internal service for listening to multipath events:
+
+- Code interested in receiving multipath events will subscribe a callback to receive the event
+- When an event is received, all subscribers will receive the events (in another thread).
+- If nobody is registered for events, the events are dropped
+- The multipath event monitor will be started when connecting to the pool
+- The multipath event monitor will be stopped when disconnecting from the pool
+
+
+A Udev event monitor is introduced, listening to device-mapper udev events.
+The relevant events for this feature are PATH_FAILED and PATH_REINSTATED:
+
+Variable Name: DM_ACTION
+Uevent Action(s): KOBJ_CHANGE
+Type: string
+Description:
+Value: Device-mapper specific action that caused the uevent action.
+	PATH_FAILED - A path has failed.
+	PATH_REINSTATED - A path has been reinstated.
+
+- The user can disable this feature in a specific host in Vdsm configuration file
+
+In the future, the monitor can be used for other code, for example maintaining LVM filter.
+
+### Vdsm multipath statistics handler
+
+To report multipath statistics to engine, we will add a multipath statistics handler.
+
+- The handler builds a map of all the faulty paths during startup.
+- The handler subscribes to multipath events during startup
+- The handler updates the data set of the faulty paths when its callback is called:
+    - PATH_FAILED - Add the faulty path to the map according to the device GUID
+    - PATH_REINSTATED - Remove the faulty path to the map according to the device GUID
+- Statistics are provided to engine via 'Host.getStats' verb.
+
+
+### New engine API
+
+New audit logs are added:
+- Faulty multipath paths on host "HOST_NAME" on devices: "GUID", "GUID" ... 
+- No faulty multipath paths on host "HOST_NAME"
+
+## Benefit to oVirt
+
+- The user will be able to detect storage issues faster.
+
+
+## User Experience
+
+New audit logs are added with information on the device path status.
+The logs are per host.
+Since the length of the information available in the audit log is limited,
+the full information of the failing paths will be available in the logs.
+The events are not generated on "real-time", but are based on polling of
+the statistics. Only if a change in the data received is detected, a new
+audit log will be created
+
+If a single LUN goes down on all the hosts, the user will get an alarm
+for each host in the DC.
+
+If a single path goes down on one of the hosts, the user will get an alarm
+only for this host.
+
+
+## Installation/Upgrade
+
+This feature does not affect engine or Vdsm installation or upgrades
+
+
+## User work-flows
+
+- Once a user connects to a new ISCSI server, no events will be created.
+
+- If there is an issue with the connection to the storage server, and the 
+multipath reports an issue with one or more device path, a "Faulty multipath paths on host" will be created
+
```

Comment:
I don't think we will get events in connection flow,  no need to mention connection.

Reply:
Done